Parsing parameter overhaul

GiudGiud opened the pull request:
This is not necessarily the right names, but I d like to start the discussion with a working PR to show the extent of the effort required for this parameter rename. 

The motivation for this is that the framework uses different types for parsed inputs and the parameters are not clear as they currently stand. With a future uniformisation of parsed capabilities + adding parsed functors and functor arguments, this will only get worse. 

But the problem is that renaming them seems extremely disruptive to the phase field module

 thoughts about this
@dschwen @jiangwen84 

Closes #19119

GiudGiud committed:
Fixup parameters for parsed system, more consistency, more framework-ey, refs #21244

GiudGiud committed:
Adjust test syntax for new parsed expression syntax, refs #20535

GiudGiud committed:
Adapt phase field parsed CH source params
Adapt modules tests to new parsed syntax, refs #20535

GiudGiud committed:
Adapt phase field parsed CH source params, refs #19119
args -> variables_names then amended to coupled_variables

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1077896/) on 1e9dca1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8857424f6409baa438ce06b4d4a9fa31f21a7b8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1077929/) on bc433a5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8857424f6409baa438ce06b4d4a9fa31f21a7b8`

dschwen commented:
:-/

Lots of changes files. Does this change the args parameter everywhere? Is variable_names the best alternative? I suppose at least for parsed materials it mirrors material property names.

GiudGiud commented:
This PR does change all the input files from args to variable_names.
I m not sure if variable_names is the best name. It s the best for ParsedMaterial for sure and that's where I determined it. 

In phase_field module, I think coupled_variables or parsed_variables is a slightly better name.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1077981/) on 6c07c2f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21296/site)

This comment will be updated on new commits.

hugary1995 commented:
This discussion has actually happened more than once before. From what I remember, we couldn't find a much better alternative for `args`. I guess we didn't see what's particularly bad about `args`, hence we didn't have enough motivation to change this parameter.

joshuahansel commented:
I've also been confused by parsed object param names. If it's open season on them, I'll suggest:
- `args` to `variable_names`
- `f_name` to `name` (unfortunately there is `material_property_names`; otherwise would call it `material_property_name`)
- `constant_expressions` to `constant_values` (why "expressions"? As I understand it, these would require `${fparse blah}` in expressions in the input file, so that they get parsed in the input file, not in the object, right?)
- `function` to `expression` (to distinguish from MOOSE Function)
- `value` to `expression`
- `vars`/`vals` is probably fine already, but maybe `expression_vars`/`expression_vals`, since `vars` can seem like coupled variables.

dschwen commented:
> * `constant_expressions` to `constant_values` (why "expressions"? As I understand it, these would require `${fparse blah}` in expressions in the input file, so that they get parsed in the input file, not in the object, right?)

No, you can use _expressions_. This is also explained in the manual. Your constant expressions can use all constants declared prior in the same parameter vector!

dschwen commented:
> * `f_name` to `name` (unfortunately there is `material_property_names`; otherwise would call it `material_property_name`)

The `f_name` stands for _function_ name. That's the name of the base material property only. In addition the material creates properties that are the derivatives, an those have generated name according to a fixed schema. I think simply calling this "material_property" is confusing.

dschwen commented:
> * `vars`/`vals` is probably fine already, but maybe `expression_vars`/`expression_vals`, since `vars` can seem like coupled variables.

vars vals appears in `ParsedFunction`. I've always found that super confusing and just prepending `expression` does not clear anything up for me.

joshuahansel commented:
> The f_name stands for function name. That's the name of the base material property only. In addition the material creates properties that are the derivatives, an those have generated name according to a fixed schema. I think simply calling this "material_property" is confusing.

"function name" seems confusing too. In light of the multiple properties, maybe `base_name`.

> vars vals appears in ParsedFunction. I've always found that super confusing and just prepending expression does not clear anything up for me.

Yeah, I was trying to communicate that `vars` are some tokens that appear in `expression` - they aren't MOOSE variables.

I don't feel strongly about any of my suggestions though, so I think I'd be happy as long as we're consistent, and the parameter descriptions no longer mention free energy 😛

GiudGiud commented:
I took care of this last thing.
I do think if we rename parameters we should do it in one go, not make multiple PRs. 
`vars` is problematic I agree

`function_name` is an issue when the parsed object may depend on a function and have a `function_names` parameter

GiudGiud commented:
how about constant_args instead of vars?

GiudGiud commented:
miscellaneous_args ?

GiudGiud commented:
This is still useful, though likely needs to be reduced in scope to be accepted

GiudGiud commented:
how about `scalar_args / scalar_vals` instead of vars / vals?
for `constants, functions, PPs, and scalar variables`

@dschwen @joshuahansel

GiudGiud commented:
so
args -> variable_names
f_name -> property_name (for ParsedMaterial) ?
constant_expressions still same
function -> expression
value -> expression
vars/vals ->scalar_vars/vals

GiudGiud commented:
umm scalar_vars still isnt great for functions. Works for PPs, scalar variables and constants

GiudGiud commented:
vars / vals
could be:
function_args (for vars)
argument_vals (for vals)

joshuahansel commented:
> so
args -> variable_names
f_name -> property_name (for ParsedMaterial) base_name ?? ?
constant_expressions still same
function -> expression
value -> expression
vars/vals ->scalar_vars/vals

I like most of these. `args` could maybe just be `variables`, but `variable_names` works too. I'm still not sure about `f_name` - even though it's technically a sort of base name, I think my vote would be `property_name`. For `vars`/`vals`, I maybe like `args`/`vals`, (I think these parameters do not exist alongside the existing `args` parameter).

dschwen commented:
`base_name` already has a different meaning and should not be used here. Anything with `scalar` screams `ScalarKernel` etc. and would be confusing, too. I do like `expression`, and for `args` I suggest `coupled_variables` (keep in mind that we use `args` in many places, among them Kernels and having both `variable` and `variables` as parameters would be ... suboptimal).

GiudGiud commented:
I m worried about re-using args for vars/vals. I think re-using a deprecated param will be headaches. 
How about expression_args expression_args/vals? or arguments/values ?

```
args -> coupled_variables
f_name -> property_name (for ParsedMaterial)
constant_expressions still same
function -> expression
value -> expression
vars/vals -> args/vals

joshuahansel commented:
I think we're close. Some other ideas:

- `expression`, `expression_args`/`expression_vals`
- `parsed_expression`, `parsed_args`/`parsed_vals`

I think `expression_vars` or `parsed_vars` would be fine too.

GiudGiud commented:
I think it s poll time?

dschwen commented:
Since the discussion apparently got split over github and slack, here's my slack comment.

Daniel Schwen
  [4 days ago](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1666127570552439)
`vars` and `vals` are a ParsedFunction thing, and `f_name` is a `ParsedMaterial thing`. I suggested `coupled_variables` to replace `args`. For `ParsedFunction` all suggestions that retain `*vals` and `*vars` in the name are not an improvement, as they clarify nothing. (edited) 

Daniel Schwen
  [4 days ago](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1666127769880799?thread_ts=1666127570.552439&cid=C01054VRUEM)
`symbol_names`, `symbol_values`

GiudGiud commented:
Current list
```
args -> coupled_variables
f_name -> property_name (for ParsedMaterial)
constant_expressions still same
material_property_names still same
function -> expression
value -> expression
vars/vals -> symbol_names/symbol_values
```

GiudGiud committed:
Change parsed function parameters:
value -> expression
vars/vals -> symbol_names symbol_values
refs #19119

GiudGiud committed:
Properly deprecate value parameter in ParsedFunction
refs #19119

GiudGiud committed:
Replace ParsedFunction vars by symbol_names in all moose/
refs #19119

GiudGiud committed:
Rename vals to symbol_values in ParsedFunction in all moose/
refs #19119

GiudGiud committed:
Better handling of deprecated parameters
Update ParsedAux to new parameters
refs #19119

GiudGiud committed:
Rename value -> expression in ParsedFunction in all moose/
refs #19119

GiudGiud committed:
Rename args to coupled_variables in all ParsedMaterials in all moose/, refs #19119

GiudGiud committed:
Rename function to expression in all ParsedMaterials in all moose/, refs #19119

GiudGiud committed:
Rename ParsedAux parameters in all moose/ inputs. refs #19119
- function -> expression
- variable_names -> coupled_variables

GiudGiud committed:
Rename args to coupled_variables in all ParsedMaterials in modules, refs #19119

GiudGiud committed:
Revert f_name -> property_name in phase field
Update (previous change from args) variable_names -> coupled_variables

refs #19119

GiudGiud committed:
Missed parameter changes refs #19119

GiudGiud committed:
Fix two heavy FV tests, refs #20953

GiudGiud committed:
Increase min AD size for mortar heavy tests that fail at 53, refs #20175

GiudGiud committed:
Adjust documentation to new parameters, refs #19119

GiudGiud committed:
Allow a heavy heat conduction test to find combined executable, refs #15595

GiudGiud committed:
Rename parameters for parsedODEKernel, refs #19119
function -> expression

GiudGiud commented:
The new params added to the scope of changes made this an even bigger PR but it should be getting close to ready now

Other changes:
- The deprecation should be handled much more smoothly than in the initial version
- Fixed a couple heavy tests as I was running the test suite locally. It s odd that CIVET doesnt run them, I think the requisites are specific enough
- I tried to avoid changing all of phase field parameters this time. AC/CH etc remain with f_name

GiudGiud commented:
#20535 was the wrong issue to use.
I ll change it to https://github.com/idaholab/moose/issues/19119 once this is full green

GiudGiud committed:
Use correct issue number for deprecation TODO, refs #19119

GiudGiud commented:
Alright all green now.
I went for `coupled_variables` instead of `args`. Changing to `variable_names` is one `sed` away so if there are second thoughts, don't hold back.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1225089/) on 61421e4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ebf4</th>
      <th colspan=4>#21296 61421e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.36%</td>
      <td>84.37%</td>
      <td>+0.01%</td>
      <td>95.79%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80674</td>
      <td>80732</td>
      <td>+58</td>
      <td>91</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14956</td>
      <td>14954</td>
      <td>-2</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21296/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21296/coverage/framework)


## Modules coverage

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ebf4</th>
      <th colspan=4>#21296 61421e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.01%</td>
      <td>89.98%</td>
      <td>-0.03%</td>
      <td>82.54%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12487</td>
      <td>12512</td>
      <td>+25</td>
      <td>52</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1386</td>
      <td>1394</td>
      <td>+8</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21296/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21296/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21296/coverage/xfem)

</details>

## Warnings

- `phase_field` new line coverage rate 82.54% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add routine to get parameters that were renamed
Handles both the case where they are missing and the case where both are provided

GiudGiud commented:
Added a getRenamedParam routine that comes in pretty handy for handling deprecation as a suggestion from @dschwen

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1278749/) on 5ec6864 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21296/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6277edf40f6afceec6dc3f80583b4152018d724b`

GiudGiud committed:
Add logic to getRenamedParam to handle default values gracefully: order of priority:
- new param value set by user
- old param value set by user
- new param default

Remove deprecated value_xyz syntax for ParsedGradFunction
This removal was at first necessary because the getRenamedParam syntax will trigger an error when the new param uses the default (I did not add the default) and the old param is specified.

Also:
- move some coord_type from Problem to Mesh to remove deprecation warning

GiudGiud commented:
should be ready @dschwen

GiudGiud commented:
@dschwen

GiudGiud commented:
Thanks!

gambka commented:
I'm working on addressing the deprecation failures related to these changes in some downstream apps. Should the deprecation of `function` in `ParsedPostprocessor` in favor of `expression` also be made?

GiudGiud commented:
There should be no failures, just warnings. 
It could be done. There wasnt as much confusion there.

gambka commented:
> There should be no failures, just warnings. It could be done. There wasnt as much confusion there.

By failure I mean we have an allowed to fail test target that runs with `--error-deprecated`. When it is no longer green I prefer to make the necessary changes earlier rather than later to avoid users copying/using inputs still using the older syntax.

lindsayad commented on modules/phase_field/src/kernels/ADACInterface.C:
```diff
@@ -18,7 +18,12 @@ ADACInterface::validParams()
   params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
   params.addParam<MaterialPropertyName>("kappa_name", "kappa_op", "The kappa used with the kernel");
-  params.addCoupledVar("args", "Vector of nonlinear variable arguments this object depends on");
+  params.addDeprecatedCoupledVar("args",
+                                 "Vector of nonlinear variable arguments this object depends on",
+                                 "args is deprecated, use 'coupled_variables' instead");
```
Oh man this is what I get for introducing an API with a default argument. This is the API:
```
  /**                                                                                                                 
   * This method adds a deprecated coupled variable name pair.  The parser will look for variable                     
   * name pair in the input file and can return a reference to the storage location                                   
   * for the coupled variable if found. The doc string for the deprecated variable will be                            
   * constructed from the doc string for the new variable. A deprecation message will also be                         
   * automatically generated                                                                                          
   */                                                                                                                 
  void addDeprecatedCoupledVar(const std::string & old_name,                                                          
                               const std::string & new_name,                                                          
                               const std::string & removal_date = "");
```
This PR introduces ~40 wrong uses

lindsayad replied:
I will be deprecating this deprecation API

GiudGiud replied:
sounds good. 
Hope my PR isnt causing too many troubles
